src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -209,6 +213,40 @@ public IndexFeatureSet indexFeatureSet()
         return indexFeatureSet;
     }
 
+    /**
+     * Returns whether this query is selection the {@link PrimaryKey}.
+     * The query selects the key if:
+     *  1. The query is not row-aware
+     *  2. The table associated with the query is not using clustering keys
+     *  3. The clustering index filter for the command wants the row.
+     *
+     *  Item 3 is important in paged queries where the {@link org.apache.cassandra.db.filter.ClusteringIndexSliceFilter} for
+     *  subsequent paged queries may not select rows that are returned by the index
+     *  search because that is initially partition based.
+     *
+     * @param key The {@link PrimaryKey} to be tested
+     * @return true if the key is selected by the query
+     */
+    public boolean selects(PrimaryKey key)
+    {
+        return !indexFeatureSet.isRowAware() ||
```

Comment:
If the query is not row-aware, does it mean the query selects just.... any key, even if the partition (token) part of it does not match? I just wonder, is there an implied assumption somewhere that the key is already in the correct partition that matches the query predicates?